[Enhancement] aws_codepipeline:  support for `Compute` action

tabito-hara opened the pull request:
### Description

* Add several arguments to the `action` block to support the `Compute` action:

  * `commands`
  * `output_variables`
  * `output_artifacts_for_compute_action`

    * This argument does not directly correspond to the AWS API objects.
    * In the AWS API, [OutputArtifacts](https://docs.aws.amazon.com/codepipeline/latest/APIReference/API_OutputArtifact.html) is an object with two attributes. However, in Terraform, the existing `output_artifacts` is implemented as a list of artifact names.
    * The `Compute` action uses the `files` attribute in `OutputArtifacts`, which cannot be represented using the current `output_artifacts`.
    * To avoid affecting existing configurations, a new argument, `output_artifacts_for_compute_action`, has been introduced. Its schema matches that of the `OutputArtifacts` object in the AWS API.
    * When the action is `Compute`, `output_artifacts_for_compute_action` is used and `output_artifacts` is ignored.
    * For all other actions, the reverse applies.

* This PR is based on #41034, rebased onto the latest `main` branch, and adds the `output_variables` and `output_artifacts_for_compute_action` arguments.

### Relations

Closes #39743
Relates #41034
Closes #42997

### References
https://docs.aws.amazon.com/codepipeline/latest/APIReference/API_ActionDeclaration.html#CodePipeline-Type-ActionDeclaration-commands
https://docs.aws.amazon.com/codepipeline/latest/userguide/action-reference-Commands.html

### Output from Acceptance Testing
Rerun after rebasing on the latest branch (2025-10-02)

```console
$ make testacc TESTS='TestAccCodePipeline_' PKG=codepipeline
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
make: Running acceptance tests on branch: 🌿 f-aws_codepipeline-add_commands 🌿...
TF_ACC=1 go1.24.6 test ./internal/service/codepipeline/... -v -count 1 -parallel 20 -run='TestAccCodePipeline_'  -timeout 360m -vet=off
2025/10/02 23:35:05 Creating Terraform AWS Provider (SDKv2-style)...
2025/10/02 23:35:06 Initializing Terraform AWS Provider (SDKv2-style)...
=== RUN   TestAccCodePipeline_basic
=== PAUSE TestAccCodePipeline_basic
=== RUN   TestAccCodePipeline_disappears
=== PAUSE TestAccCodePipeline_disappears
=== RUN   TestAccCodePipeline_emptyStageArtifacts
=== PAUSE TestAccCodePipeline_emptyStageArtifacts
=== RUN   TestAccCodePipeline_deployWithServiceRole
=== PAUSE TestAccCodePipeline_deployWithServiceRole
=== RUN   TestAccCodePipeline_tags
=== PAUSE TestAccCodePipeline_tags
=== RUN   TestAccCodePipeline_MultiRegion_basic
=== PAUSE TestAccCodePipeline_MultiRegion_basic
=== RUN   TestAccCodePipeline_MultiRegion_update
=== PAUSE TestAccCodePipeline_MultiRegion_update
=== RUN   TestAccCodePipeline_MultiRegion_convertSingleRegion
=== PAUSE TestAccCodePipeline_MultiRegion_convertSingleRegion
=== RUN   TestAccCodePipeline_withNamespace
=== PAUSE TestAccCodePipeline_withNamespace
=== RUN   TestAccCodePipeline_withGitHubV1SourceAction
    codepipeline_test.go:492: Environment variable GITHUB_TOKEN is not set, skipping test
--- SKIP: TestAccCodePipeline_withGitHubV1SourceAction (0.00s)
=== RUN   TestAccCodePipeline_ecr
=== PAUSE TestAccCodePipeline_ecr
=== RUN   TestAccCodePipeline_pipelineType
=== PAUSE TestAccCodePipeline_pipelineType
=== RUN   TestAccCodePipeline_manualApprovalTimeoutInMinutes
=== PAUSE TestAccCodePipeline_manualApprovalTimeoutInMinutes
=== RUN   TestAccCodePipeline_conditions
=== PAUSE TestAccCodePipeline_conditions
=== RUN   TestAccCodePipeline_computeAction
=== PAUSE TestAccCodePipeline_computeAction
=== RUN   TestAccCodePipeline_trigger
=== PAUSE TestAccCodePipeline_trigger
=== CONT  TestAccCodePipeline_basic
=== CONT  TestAccCodePipeline_withNamespace
=== CONT  TestAccCodePipeline_conditions
=== CONT  TestAccCodePipeline_trigger
=== CONT  TestAccCodePipeline_computeAction
=== CONT  TestAccCodePipeline_pipelineType
=== CONT  TestAccCodePipeline_manualApprovalTimeoutInMinutes
=== CONT  TestAccCodePipeline_tags
=== CONT  TestAccCodePipeline_MultiRegion_convertSingleRegion
=== CONT  TestAccCodePipeline_MultiRegion_update
=== CONT  TestAccCodePipeline_MultiRegion_basic
=== CONT  TestAccCodePipeline_emptyStageArtifacts
=== CONT  TestAccCodePipeline_deployWithServiceRole
=== CONT  TestAccCodePipeline_disappears
=== CONT  TestAccCodePipeline_ecr
--- PASS: TestAccCodePipeline_deployWithServiceRole (52.31s)
--- PASS: TestAccCodePipeline_computeAction (58.95s)
--- PASS: TestAccCodePipeline_disappears (59.18s)
--- PASS: TestAccCodePipeline_emptyStageArtifacts (59.97s)
--- PASS: TestAccCodePipeline_withNamespace (67.88s)
--- PASS: TestAccCodePipeline_ecr (72.42s)
--- PASS: TestAccCodePipeline_trigger (75.23s)
--- PASS: TestAccCodePipeline_MultiRegion_basic (78.39s)
--- PASS: TestAccCodePipeline_conditions (85.66s)
--- PASS: TestAccCodePipeline_basic (87.62s)
--- PASS: TestAccCodePipeline_manualApprovalTimeoutInMinutes (89.55s)
--- PASS: TestAccCodePipeline_tags (102.31s)
--- PASS: TestAccCodePipeline_MultiRegion_update (103.87s)
--- PASS: TestAccCodePipeline_MultiRegion_convertSingleRegion (111.02s)
--- PASS: TestAccCodePipeline_pipelineType (145.75s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/codepipeline       150.697s



```

tabito-hara committed:
add commands and output_artifacts_for_compute_action

tabito-hara committed:
add output_variables argument, along with its acctest

tabito-hara committed:
update documentation

tabito-hara committed:
add changelog

tabito-hara committed:
simplified

Caesta-Altostratus commented:
Any news on this?

jatin-mehrotra-colorkrew commented:
When will this be available? I am also waiting on this.

tabito-hara commented:
Rebased on the latest main branch

adamsebesta commented:
Waiting on this one too!

robinsq commented:
@ewbankkit - this functionality has been out since Sept 2024 - may we get it in terraform?

robinsq commented:
@jar-b @YakDriver @ewbankkit  - I pinged you all because you all have recently approved/merged PRs for this repository.

May we have this PR reviewed so that we can gain this AWS Code Pipeline functionality to be managed via code.
The feature has been released in 2024.

robinsq commented:
:wave: @johnsonaj  Adrian, I recently saw that you merged a PR on this repository.

This AWS Codepipeline has been released back in 2024. We should upgrade the provider to support it.
The community has provided a PR for this feature and it has been reviewed by a dev (not in code owners).

May we have it reviewed and merged?

robinsq commented:
@hashicorp/terraform-aws  - anything else the community can do to help move this forward?

lmayorga1980 commented:
This is already in CF, why not in TF.

subham-ibmhc commented on internal/service/codepipeline/codepipeline.go:
```diff
@@ -485,6 +494,40 @@ func resourcePipeline() *schema.Resource {
 											Optional: true,
 											Elem:     &schema.Schema{Type: schema.TypeString},
 										},
+										"output_artifacts_for_compute_action": {
```
copilot-feedback:actionable

subham-ibmhc commented on internal/service/codepipeline/codepipeline.go:
```diff
@@ -1025,6 +1068,14 @@ func expandActionDeclaration(tfMap map[string]any) *types.ActionDeclaration {
 		apiObject.ActionTypeId.Category = types.ActionCategory(v)
 	}
 
+	if v, ok := tfMap["commands"].([]any); ok && len(v) > 0 {
+		for _, v := range v {
+			if v, ok := v.(string); ok && v != "" {
+				apiObject.Commands = append(apiObject.Commands, v)
+			}
+		}
+	}
```
copilot-feedback:noise

subham-ibmhc commented on internal/service/codepipeline/codepipeline.go:
```diff
@@ -1041,8 +1092,22 @@ func expandActionDeclaration(tfMap map[string]any) *types.ActionDeclaration {
+		}
+	}
+
+	if v, ok := tfMap["output_variables"].([]any); ok && len(v) > 0 {
+		for _, v := range v {
+			if v, ok := v.(string); ok && v != "" {
+				apiObject.OutputVariables = append(apiObject.OutputVariables, v)
+			}
+		}
 	}
```
copilot-feedback:noise

subham-ibmhc commented on internal/service/codepipeline/codepipeline.go:
```diff
@@ -1988,6 +2108,29 @@ func flattenOutputArtifacts(apiObjects []types.OutputArtifact) []string {
+		}
+		if v := apiObject.Files; len(v) > 0 {
+			tfMap["files"] = v
+		} else {
+			tfMap["files"] = nil
+		}
```
copilot-feedback:actionable

subham-ibmhc commented on website/docs/r/codepipeline.html.markdown:
```diff
@@ -223,9 +223,14 @@ An `action` block supports the following arguments:
 * `name` - (Required) The action declaration's name.
 * `provider` - (Required) The provider of the service being called by the action. Valid providers are determined by the action category. Provider names are listed in the [Action Structure Reference](https://docs.aws.amazon.com/codepipeline/latest/userguide/action-reference.html) documentation.
 * `version` - (Required) A string that identifies the action type.
+* `commands` - (Optional) A list of shell commands to run with the compute action.
```
copilot-feedback:actionable

subham-ibmhc commented on website/docs/r/codepipeline.html.markdown:
```diff
@@ -223,9 +223,14 @@ An `action` block supports the following arguments:
 * `input_artifacts` - (Optional) A list of artifact names to be worked on.
-* `output_artifacts` - (Optional) A list of artifact names to output. Output artifact names must be unique within a pipeline.
+* `output_artifacts` - (Optional) A list of artifact names to output. Output artifact names must be unique within a pipeline. If the action is `Compute`, this argument is ignored.
+* `output_artifacts_for_compute_action` - (Optional) A block of output artifacts for the compute action. If the action is not `Compute`, this argument is ignored.
```
copilot-feedback:actionable

subham-ibmhc commented on internal/service/codepipeline/codepipeline.go:
```diff
@@ -485,6 +494,40 @@ func resourcePipeline() *schema.Resource {
 											Optional: true,
 											Elem:     &schema.Schema{Type: schema.TypeString},
 										},
+										"output_artifacts_for_compute_action": {
+											Type:     schema.TypeList,
+											Optional: true,
+											Elem: &schema.Resource{
+												Schema: map[string]*schema.Schema{
```
copilot-feedback:actionable

subham-ibmhc committed:
fix initial tests

subham-ibmhc committed:
Add customizeDiff to validate edge cases

subham-ibmhc committed:
Fix documentation

subham-ibmhc committed:
fix test

subham-ibmhc commented:
```console
% make t T=TestAccCodePipeline_ K=codepipeline 
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
make: Running acceptance tests on branch: 🌿 f-aws_codepipeline-add_commands 🌿...
TF_ACC=1 go1.26.5 test ./internal/service/codepipeline/... -v -count 1 -parallel 20 -run='TestAccCodePipeline_'  -timeout 360m -vet=off
2026/07/21 17:35:57 Creating Terraform AWS Provider (SDKv2-style)...
2026/07/21 17:35:57 Initializing Terraform AWS Provider (SDKv2-style)...
=== RUN   TestAccCodePipeline_List_basic
=== PAUSE TestAccCodePipeline_List_basic
=== RUN   TestAccCodePipeline_List_includeResource
=== PAUSE TestAccCodePipeline_List_includeResource
=== RUN   TestAccCodePipeline_List_regionOverride
=== PAUSE TestAccCodePipeline_List_regionOverride
=== RUN   TestAccCodePipeline_basic
=== PAUSE TestAccCodePipeline_basic
=== RUN   TestAccCodePipeline_disappears
=== PAUSE TestAccCodePipeline_disappears
=== RUN   TestAccCodePipeline_emptyStageArtifacts
=== PAUSE TestAccCodePipeline_emptyStageArtifacts
=== RUN   TestAccCodePipeline_deployWithServiceRole
=== PAUSE TestAccCodePipeline_deployWithServiceRole
=== RUN   TestAccCodePipeline_tags
=== PAUSE TestAccCodePipeline_tags
=== RUN   TestAccCodePipeline_MultiRegion_basic
=== PAUSE TestAccCodePipeline_MultiRegion_basic
=== RUN   TestAccCodePipeline_MultiRegion_update
=== PAUSE TestAccCodePipeline_MultiRegion_update
=== RUN   TestAccCodePipeline_MultiRegion_convertSingleRegion
=== PAUSE TestAccCodePipeline_MultiRegion_convertSingleRegion
=== RUN   TestAccCodePipeline_withNamespace
=== PAUSE TestAccCodePipeline_withNamespace
=== RUN   TestAccCodePipeline_withGitHubV1SourceAction
    codepipeline_test.go:499: skipping test; environment variable GITHUB_TOKEN must be set
--- SKIP: TestAccCodePipeline_withGitHubV1SourceAction (0.00s)
=== RUN   TestAccCodePipeline_ecr
=== PAUSE TestAccCodePipeline_ecr
=== RUN   TestAccCodePipeline_pipelineType
=== PAUSE TestAccCodePipeline_pipelineType
=== RUN   TestAccCodePipeline_manualApprovalTimeoutInMinutes
=== PAUSE TestAccCodePipeline_manualApprovalTimeoutInMinutes
=== RUN   TestAccCodePipeline_conditions
=== PAUSE TestAccCodePipeline_conditions
=== RUN   TestAccCodePipeline_computeAction
=== PAUSE TestAccCodePipeline_computeAction
=== RUN   TestAccCodePipeline_trigger
=== PAUSE TestAccCodePipeline_trigger
=== CONT  TestAccCodePipeline_List_basic
=== CONT  TestAccCodePipeline_MultiRegion_update
=== CONT  TestAccCodePipeline_manualApprovalTimeoutInMinutes
=== CONT  TestAccCodePipeline_ecr
=== CONT  TestAccCodePipeline_pipelineType
=== CONT  TestAccCodePipeline_withNamespace
=== CONT  TestAccCodePipeline_tags
=== CONT  TestAccCodePipeline_computeAction
=== CONT  TestAccCodePipeline_MultiRegion_basic
=== CONT  TestAccCodePipeline_trigger
=== CONT  TestAccCodePipeline_disappears
=== CONT  TestAccCodePipeline_conditions
=== CONT  TestAccCodePipeline_emptyStageArtifacts
=== CONT  TestAccCodePipeline_deployWithServiceRole
=== CONT  TestAccCodePipeline_MultiRegion_convertSingleRegion
=== CONT  TestAccCodePipeline_basic
=== CONT  TestAccCodePipeline_List_regionOverride
=== CONT  TestAccCodePipeline_List_includeResource
=== NAME  TestAccCodePipeline_emptyStageArtifacts
    codepipeline_test.go:179: Step 1/2 error: Error running apply: exit status 1
        
        Error: creating S3 Bucket (tf-test-pipeline-test-d4lozsz4fd): operation error S3: CreateBucket, https response error StatusCode: 0, RequestID: , HostID: , request send failed, Put "https://tf-test-pipeline-test-d4lozsz4fd.s3.us-west-2.amazonaws.com/": dial tcp: lookup tf-test-pipeline-test-d4lozsz4fd.s3.us-west-2.amazonaws.com: no such host
        
          with aws_s3_bucket.test,
          on terraform_plugin_test.tf line 12, in resource "aws_s3_bucket" "test":
          12: resource "aws_s3_bucket" "test" {
        
--- FAIL: TestAccCodePipeline_emptyStageArtifacts (20.31s)
--- PASS: TestAccCodePipeline_List_regionOverride (46.99s)
--- PASS: TestAccCodePipeline_disappears (58.23s)
--- PASS: TestAccCodePipeline_List_basic (59.31s)
--- PASS: TestAccCodePipeline_withNamespace (60.66s)
--- PASS: TestAccCodePipeline_computeAction (62.43s)
--- PASS: TestAccCodePipeline_MultiRegion_basic (65.12s)
--- PASS: TestAccCodePipeline_List_includeResource (67.98s)
--- PASS: TestAccCodePipeline_ecr (72.70s)
--- PASS: TestAccCodePipeline_deployWithServiceRole (75.84s)
--- PASS: TestAccCodePipeline_basic (95.95s)
--- PASS: TestAccCodePipeline_trigger (96.76s)
--- PASS: TestAccCodePipeline_manualApprovalTimeoutInMinutes (99.06s)
--- PASS: TestAccCodePipeline_MultiRegion_update (100.19s)
--- PASS: TestAccCodePipeline_conditions (103.06s)
--- PASS: TestAccCodePipeline_tags (113.87s)
--- PASS: TestAccCodePipeline_MultiRegion_convertSingleRegion (131.59s)
--- PASS: TestAccCodePipeline_pipelineType (183.52s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-aws/internal/service/codepipeline       188.547s
FAIL
make: *** [t] Error 1
subhammukhopadhyay@Subhams-MacBook-Pro terraform-provider-aws % make t T=TestAccCodePipeline_emptyStageArtifacts K=codepipeline
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
make: Running acceptance tests on branch: 🌿 f-aws_codepipeline-add_commands 🌿...
TF_ACC=1 go1.26.5 test ./internal/service/codepipeline/... -v -count 1 -parallel 20 -run='TestAccCodePipeline_emptyStageArtifacts'  -timeout 360m -vet=off
2026/07/21 17:52:39 Creating Terraform AWS Provider (SDKv2-style)...
2026/07/21 17:52:39 Initializing Terraform AWS Provider (SDKv2-style)...
=== RUN   TestAccCodePipeline_emptyStageArtifacts
=== PAUSE TestAccCodePipeline_emptyStageArtifacts
=== CONT  TestAccCodePipeline_emptyStageArtifacts
--- PASS: TestAccCodePipeline_emptyStageArtifacts (44.73s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/codepipeline       49.696s
```

jar-b commented on internal/service/codepipeline/codepipeline.go:
```diff
@@ -2205,3 +2347,26 @@ func flattenTriggerDeclarations(apiObjects []types.PipelineTriggerDeclaration) [
 	return tfList
 }
+
+func outputArtifactsConflictValidate(_ context.Context, diff *schema.ResourceDiff, _ any) error {
```
It may be worth adding a unit test for this, just due to the level of nesting and various attribute combinations that could trigger an error.

Alternatively, an acceptance test with invalid argument combinations also could work.

subham-ibmhc replied:
Added acceptance tests to check the validation.

jar-b commented on .changelog/42507.txt:
````diff
@@ -0,0 +1,3 @@
+```release-note:enhancement
+resource/aws_codepipeline: Support for `Compute` action
````
nit - explicitly state the new argument names (`commands` and `output_artifacts_for_compute_action`).

It might also be worth a second entry that states `output_artifacts` and `output_artifacts_for_compute_action` now conflict.

subham-ibmhc committed:
Add tests for customizeDiff edge cases

```console
% make t T=TestAccCodePipeline_nonComputeActionWithOutputAttributesForComputeAction K=codepipeline
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
make: Running acceptance tests on branch: 🌿 f-aws_codepipeline-add_commands 🌿...
TF_ACC=1 go1.26.5 test ./internal/service/codepipeline/... -v -count 1 -parallel 20 -run='TestAccCodePipeline_nonComputeActionWithOutputAttributesForComputeAction'  -timeout 360m -vet=off
2026/07/22 15:20:31 Creating Terraform AWS Provider (SDKv2-style)...
2026/07/22 15:20:31 Initializing Terraform AWS Provider (SDKv2-style)...
=== RUN   TestAccCodePipeline_nonComputeActionWithOutputAttributesForComputeAction
=== PAUSE TestAccCodePipeline_nonComputeActionWithOutputAttributesForComputeAction
=== CONT  TestAccCodePipeline_nonComputeActionWithOutputAttributesForComputeAction
--- PASS: TestAccCodePipeline_nonComputeActionWithOutputAttributesForComputeAction (5.77s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/codepipeline       10.733s
subhammukhopadhyay@Subhams-MacBook-Pro terraform-provider-aws % make t T=TestAccCodePipeline_computeActionWithOutputAttributes K=codepipeline
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
make: Running acceptance tests on branch: 🌿 f-aws_codepipeline-add_commands 🌿...
TF_ACC=1 go1.26.5 test ./internal/service/codepipeline/... -v -count 1 -parallel 20 -run='TestAccCodePipeline_computeActionWithOutputAttributes'  -timeout 360m -vet=off
2026/07/22 15:23:19 Creating Terraform AWS Provider (SDKv2-style)...
2026/07/22 15:23:19 Initializing Terraform AWS Provider (SDKv2-style)...
=== RUN   TestAccCodePipeline_computeActionWithOutputAttributes
=== PAUSE TestAccCodePipeline_computeActionWithOutputAttributes
=== CONT  TestAccCodePipeline_computeActionWithOutputAttributes
--- PASS: TestAccCodePipeline_computeActionWithOutputAttributes (5.39s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/codepipeline       10.344s
```

subham-ibmhc committed:
changelog update

subham-ibmhc committed:
terrafmt

jar-b commented on .changelog/42507.txt:
````diff
@@ -0,0 +1,7 @@
+```release-note:enhancement
+resource/aws_codepipeline: Added new arguments `commands` and 'output_artifacts_for_compute_action' to support 'Compute' action
````
```suggestion
resource/aws_codepipeline: Add `stage.action.commands` and `stage.action.output_artifacts_for_compute_action` arguments to support Compute action types
```

jar-b commented on .changelog/42507.txt:
````diff
@@ -0,0 +1,7 @@
+```
+
+```release-note:enhancement
+resource/aws_codepipeline: 'output_artifacts_for_compute_action' and 'output_artifacts' now conflict
````
```suggestion
resource/aws_codepipeline: `stage.action.output_artifacts_for_compute_action` and `stage.action.output_artifacts` now conflict
```

subham-ibmhc committed:
Update .changelog/42507.txt

Co-authored-by: Jared Baker <jar-b@users.noreply.github.com>

subham-ibmhc committed:
Update .changelog/42507.txt

Co-authored-by: Jared Baker <jar-b@users.noreply.github.com>

jar-b reviewed (approved):
LGTM 🚀
